### PR TITLE
Support NtQuerySystemInformation API SystemBasicInformation

### PR DESCRIPTION
**Pull request by nghiadt1098**

### CODE REVIEW

**Comment by nghiadt1098** [comment]
Hi I've implemented ntdll!NtQuerySystemInformation for X8664 windows with SystemInformationClass is SystemBasicInformation.
This API is used to get PageSize, MiniumUsermodeAddress,MaxiumUsermodeAddress, Number of CPU Core,etc and used in GetSystemInfo function.

`https://docs.microsoft.com/en-us/windows/win32/api/sysinfoapi/nf-sysinfoapi-getsysteminfo`

**Comment by xwings** [comment]
Hi, 

do you have a test case this ?

**Comment by nghiadt1098** [comment]
Yes. i've prepared a small program bellow which call NtQuerySystemInformation since the _SYSTEM_BASIC_INFORMATION struct is undoccumented i must construct _SYSTEM_BASIC_INFORMATION2 struct .

```#include <windows.h>
#include <stdio.h>
#include <Winternl.h >
#include <stdlib.h>
#pragma comment(lib ,"ntdll.lib")
typedef struct _SYSTEM_BASIC_INFORMATION2
{
	ULONG Reserved;
	ULONG TimerResolution;
	ULONG PageSize;
	ULONG NumberOfPhysicalPages;
	ULONG LowestPhysicalPageNumber;
	ULONG HighestPhysicalPageNumber;
	ULONG AllocationGranularity;
	ULONG_PTR MinimumUserModeAddress;
	ULONG_PTR MaximumUserModeAddress;
	ULONG_PTR ActiveProcessorsAffinityMask;
	CCHAR NumberOfProcessors;
} SYSTEM_BASIC_INFORMATION2, * PSYSTEM_BASIC_INFORMATION2;
int main()
{
	ULONG whyme=0;
	SYSTEM_BASIC_INFORMATION2 sbi;

	NTSTATUS status = NtQuerySystemInformation(SystemBasicInformation, &sbi, sizeof(SYSTEM_BASIC_INFORMATION2), &whyme);
	printf("Size : 0x%x Status				: 0x%p\n", whyme,status);
	printf("ActiveProcessorsAffinityMask	: 0x%p\n", sbi.ActiveProcessorsAffinityMask);
	printf("AllocationGranularity			: 0x%p\n", sbi.AllocationGranularity);
	printf("HighestPhysicalPageNumber		: 0x%p\n", sbi.HighestPhysicalPageNumber);
	printf("LowestPhysicalPageNumber		: 0x%p\n", sbi.LowestPhysicalPageNumber);
	printf("MaximumUserModeAddress			: 0x%p\n", sbi.MaximumUserModeAddress);
	printf("MinimumUserModeAddress			: 0x%p\n", sbi.MinimumUserModeAddress);
	printf("NumberOfPhysicalPages			: 0x%p\n", sbi.NumberOfPhysicalPages);
	printf("NumberOfProcessors				: 0x%p\n", sbi.NumberOfProcessors);
	printf("PageSize						: 0x%p\n", sbi.PageSize);
	printf("TimerResolution					: 0x%p\n", sbi.TimerResolution);

}```

**Comment by nghiadt1098** [comment]
It's somehow quite similar as SYSTEM_INFO struct because GetSystemInfo call this syscall to get data. But of cause, target file could call nt syscall directly to gather system information. 
`https://docs.microsoft.com/en-us/windows/win32/api/sysinfoapi/ns-sysinfoapi-system_info `

**Comment by aquynh** [comment]
Nice work! You can include both source code and binary of your sample in this PR.

**Comment by nghiadt1098** [comment]
I've just commit sample for 32bit version. It's also work perfectly on 64bit version, but since python `printf` can't handle `%p`. I couldn't provide a beautiful output with `printf` the result is correct , i've dump hex the buffer.

File: qiling/os/windows/dlls/ntdll.py [hunk 1]
```diff
@@ -103,7 +103,88 @@ def hook_NtQueryInformationProcess(ql, address, params):
...
+                                                              PageSize=ql.os.heap.page_size,
+                                                              NumberOfPhysicalPages = 0x003FC38A,
+                                                              LowestPhysicalPageNumber=1,
+                                                              HighestPhysicalPageNumber=0x0046DFFF,
+                                                              AllocationGranularity=1,
+                                                              MinimumUserModeAddress=0x10000,
+                                                              MaximumUserModeAddress=0x7FFFFFFEFFFF,
+                                                              ActiveProcessorsAffinityMask = 0x3F,
+                                                              NumberOfProcessors = 0x6)
+        elif(ql.archtype == QL_ARCH.X86):
```
Comment (aquynh) [inline qiling/os/windows/dlls/ntdll.py:125]: please leave a space after `elif`

also on this line, you can remove `(` and `)`

Comment (nghiadt1098) [reply qiling/os/windows/dlls/ntdll.py]: thanks a lot for fix my coding style, do you have any reference about python coding standard. I'm a newbie in software development.

**Comment by xwings** [comment]
Hi,

there is a update with those Windows API.
Appreciate if you can update your code.

**Comment by nghiadt1098** [comment]
Hi,
i've tested NtQuerySystemInformation on Windows Preview Insider and the result is the same.
Which APIs are you mentioned?

**Comment by xwings** [comment]
Hi,

pull the latest dev code and you can see the different in our winapi implementation format.

**Comment by nghiadt1098** [comment]
Hi, 
I've some trouble with Travis CI , can you help me with this .

**Comment by xwings** [comment]
U dont have to unzip. travis will unzip. your previous error could cause by travis update. we just fixed it.

can you remove the code and try again ?

**Comment by xwings** [comment]
one last request, can u add ntQuerySystemInfo_x86.py's code to test/test_pe ?

**Comment by xwings** [comment]
Cool, looks good. I will merge now !

Once again, welcome to Qiling Framework

**Pull request merged**